### ISSUE TITLE
Recursive Folders (due to symlinks?) in Linux

### ISSUE THREAD

**Issue by mojodmonkey** [duplicati/duplicati#3536]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.4.5_beta_2018-11-28
- **Operating system**: Ubuntu 16.04.5 LTS
- **Backend**: B2 Cloud Storage

## Description

Selecting to back up all of root (/) returns a backup size that is multiple terabytes more than what the total capacity of all drives being backed up. 

## Steps to reproduce

1. Install 2.0.4.5_beta_2018-11-28 in ubuntu
2. create a backup and add a path directly of root (/) or choose a folder with a known symlink to create a recursive issue (example: "/snap/core/")
3. Run backup

- **Actual result**:
Duplicati begins backup and after a bit lists the total backup of the selected backup left in the progress bar as 640 Terabytes
- **Expected result**:
Duplicati begins back and after a bit lists the total backup of the selected backup left in the progress bar as 56 Gigabytes (what is actually selected for backup excluding the recursiveness)

## Screenshots

![](https://forum.duplicati.com/uploads/default/original/2X/3/3920adba7b51629627ab551d08ed4e15532efc49.png)

## Debug log

https://www.dropbox.com/s/nnluuwya2jzl2b7/bugreport.zip?dl=0

**Comment by warwickmm** [comment]
Are the symlink targets actually backed up, or this just an issue in the display?  I believe the default symlink policy is to just record the symlink with its name and destination.

**Comment by mojodmonkey** [comment]
I don't really know. I'm trying to create my first backup. The bigger issue is that when duplicati shows the giant size of the backup left it doesn't take long once this progress bar shows that number that my internet connection dies, and the modem requires a restart. I selected a simple directory with only 1.8GB and it uploaded just fine and didn't kill my internet. I also managed to upload all 54 GB off the main drive with the previous version, so I definitely think something strange is happening with this version compared to the previous beta.

**Comment by warwickmm** [comment]
I configured a simple test case with the following source folder:
```
source_folder/
    - 10MB_file
    - sub_folder/
        - symlink_to_source_folder
```

I then ran the backup.  It finished too quickly to see the details in the progress bar, but after the backup completed the reported size of the source files was 10 MB.  A restore was also successful in recreating the source, including the symlink.  This was with version v2.0.4.5-2.0.4.5_beta_2018-11-28 on an Arch Linux machine.

I'm not familiar with the `/snap` folder in Ubuntu.  Are there any hard links?

Also, do you have any custom configuration settings that might affect this?

**Comment by mojodmonkey** [comment]
I'm waiting on a backup of around 50GB to see if indeed the backup will respond back as successful or simply go beyond 50GB (which is how I'll know if duplicati also backs up unnecessary files)

**Comment by mojodmonkey** [comment]
This morning the backup size was only 21GB. But the backup stopped with errors that some plex files are inconsistent. I tried reruning the backup and it immediately failed with the same issues. I tried running a db repair but am now getting the error: No filelists found on the remote destination
I've created another bug report. I don't know if that is helpful. 
https://www.dropbox.com/s/oacjxjt072s8m25/bugreport-2.zip?dl=0

**Comment by warwickmm** [comment]
If you look at your B2 storage, do you see any files with names that end with `.dlist.zip.aes`?

**Comment by mojodmonkey** [comment]
No, there is no file that ends with a .dlist.zip.aes on my b2 storage. What does that mean

**Comment by warwickmm** [comment]
Missing `dlist` files can be rebuilt from a database repair operation, as long as the local database does not have any issues.  Did the "No filelists found on the remote destination" error occur during the repair, or during a backup after the repair?

**Comment by mojodmonkey** [comment]
I got the no file list found on the remote destination error when I attempted to repair.

If I try to backup again I get:

`The database was attempted repaired, but the repair did not complete. This database may be incomplete and the backup process cannot continue. You may delete the local database and attempt to repair it again.`

If I choose repair, I get:

`No filelists found on the remote destination`

If I try Recreate (delete and repair) I get:
`No filelists found on the remote destination`

I tried to simply delete, then repair the db and got the same result.

**Comment by warwickmm** [comment]
Did an initial backup ever finish?  This could the result of an incomplete initial backup.

**Comment by mojodmonkey** [comment]
No, the initial backup never finished. I checked on the backup that I started yesterday and the backup was stopped saying that some files (plex files) were inconsist. How can I restart the backup from where it is if I can't recreate the db? Do I need to delete the backup and start from scratch?

**Comment by mojodmonkey** [comment]
Ill give that a shot when I get back home in a few days. If anyone else is familiar with this type of issue, any suggestions are appreciated.

**Comment by madmaze** [comment]
I seem to have a similar issue  (remote destination via SSH). 
It finds about 200TB worth of files (in reality its around 1.7TB), the backup does eventually finish after nearly 20hours.
I believe this issue was introduced in `2.0.4.5_beta_2018-11-28` I don't remember which version I was on before, but that that point in time the backup only took about 4hours.

**Comment by warwickmm** [comment]
@madmaze, is the size of the backup on the backend closer to 1.7 TB or 200 TB?  Also, can you provide details on the source operating system and filesystem?  It looks like others have had this issue, a few of which are using copy-on-write filesystems (btrfs, zfs) although I don't think that should matter.

https://forum.duplicati.com/t/size-to-go-is-wrong/1142
https://forum.duplicati.com/t/backup-to-go-file-size-seems-completely-wrong/1282
https://forum.duplicati.com/t/source-size-is-wrong/826
https://forum.duplicati.com/t/size-to-go-larger-than-source-size/1341

**Comment by madmaze** [comment]
@warwickmm, at the remote end its about 1.8TB, I wish I had 200TB of storage ;)

On the source side its Ubuntu 16.04, source on zfs, about 6.5m files
On the destination side its Ubuntu 18.04, target on ext4, 3 backup versions

I checked to the number of files as suggested [here](https://forum.duplicati.com/t/source-size-is-wrong/826/6) and compared it against what `find . | wc -l` returned, the numbers are a bit different, but not significantly.
Find: `7677218`
Duplicati: `7677723`

**Comment by warwickmm** [comment]
> ... at the remote end its about 1.8TB, ...

This makes me think that the issue is in how Duplicati estimates the size of the source files, and not an issue with recursive symlinks.  However, it does seem like it happens more often to those with CoW filesystems (zfs, btrfs).

@mojodmonkey, it sounds like you are encountering several other issues in addition to this one.  However, can you report back with the type of filesystem your source files reside on?

**Comment by mojodmonkey** [comment]
I'm backing up ext4

**Comment by warwickmm** [comment]
Are the source files on ext4 as well?

**Comment by mojodmonkey** [comment]
I mean that all the sources files are on ext4 being backed up tk Backblaze B2

**Comment by warwickmm** [comment]
Ah, ok.  Thanks.

**Comment by warwickmm** [comment]
Would someone be willing to try the GUI Command line option for a backup with the `dry-run`, `full-result` and `log-file` options?  Hopefully, by inspecting the list of files, you can get an idea of what may be causing the recursion.
